core/trino-main/src/main/java/io/trino/connector/system/TableCommentSystemTable.java

```diff
@@ -100,60 +94,39 @@ public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, Connect
         for (String catalog : listCatalogNames(session, metadata, accessControl, catalogFilter)) {
             QualifiedTablePrefix prefix = tablePrefix(catalog, schemaFilter, tableFilter);
 
-            Set<SchemaTableName> names = ImmutableSet.of();
-            Map<SchemaTableName, ViewInfo> views = ImmutableMap.of();
-            Map<SchemaTableName, ViewInfo> materializedViews = ImmutableMap.of();
-            try {
-                materializedViews = getMaterializedViews(session, metadata, accessControl, prefix);
-                views = getViews(session, metadata, accessControl, prefix);
-                // Some connectors like blackhole, accumulo and raptor don't return views in listTables
-                // Materialized views are consistently returned in listTables by the relevant connectors
-                names = union(listTables(session, metadata, accessControl, prefix), views.keySet());
-            }
-            catch (TrinoException e) {
-                // listTables throws an exception if cannot connect the database
-                LOG.warn(e, "Failed to get tables for catalog: %s", catalog);
-            }
+            List<TableCommentMetadata> relationComments = metadata.listRelationComments(session, prefix);
+
+            Set<SchemaTableName> accessible = accessControl.filterTables(
+                    session.toSecurityContext(),
+                    catalog,
+                    relationComments.stream()
+                            .map(TableCommentMetadata::table)
+                            .collect(toImmutableSet()));
 
-            for (SchemaTableName name : names) {
-                Optional<String> comment = Optional.empty();
-                try {
-                    comment = getComment(session, prefix, name, views, materializedViews);
+            for (TableCommentMetadata commentMetadata : relationComments) {
+                SchemaTableName name = commentMetadata.table();
+                if (!accessible.contains(name)) {
+                    continue;
                 }
-                catch (RuntimeException e) {
-                    // getTableHandle may throw an exception (e.g. Cassandra connector doesn't allow case insensitive column names)
-                    LOG.warn(e, "Failed to get metadata for table: %s", name);
+                if (!commentMetadata.redirected()) {
+                    table.addRow(catalog, name.getSchemaName(), name.getTableName(), commentMetadata.comment().orElse(null));
+                }
+                else {
+                    try {
+                        // TODO (https://github.com/trinodb/trino/issues/18514) this should consult accessControl. Leaving for now as-is.
```

Comment:
https://github.com/trinodb/trino/issues/18514.
the change here is supposed to be a refactor, so I did not attempt to fix this alleged problem